cmds/core/printf/utils.go

```diff
@@ -0,0 +1,211 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+func interpret(
+	w *bytes.Buffer,
+	format string,
+	args []string,
+	octalPrefix bool,
+	parseSubstitutions bool,
+) error {
+	o := w
+	fr := strings.NewReader(format)
+	idx := 0
+	nextArg := func() string {
+		if idx >= len(args) {
+			return ""
+		}
+		ans := args[idx]
+		idx = idx + 1
+		return ans
+	}
+
+	for fr.Len() > 0 {
+		c, _, err := fr.ReadRune()
+		// errors are only EOF
+		if err != nil {
+			continue
+		}
+		if c == '%' && parseSubstitutions {
+			// at this point we are looking for which format code this is
+			// read another rune
+			n, _, err := fr.ReadRune()
+			// error only EOF, so write the original rune and continue
+			if err != nil {
+				o.WriteRune(c)
+				continue
+			}
+			arg := nextArg()
+			switch n {
+			case '%':
+				o.WriteRune('%')
+				continue
+			case 'b':
+				tmp := &bytes.Buffer{}
+				err := interpret(tmp, arg, nil, true, false)
+				if err != nil {
+					return err
+				}
+				o.WriteString(tmp.String())
+				continue
+			case 'q':
+				continue
+			case 'd':
+				continue
+			case 'i':
+				continue
+			case 'o':
+				continue
+			case 'u':
+				continue
+			case 'x':
+				continue
+			case 'X':
+				continue
+			case 'f':
+				continue
+			case 'e':
+				continue
+			case 'E':
+				continue
+			case 'g':
+				continue
+			case 'G':
+				continue
+			case 'c':
+				continue
+			case 's':
+				fmt.Fprintf(o, "%s", arg)
+				continue
+			default:
+				return fmt.Errorf("%s: %s", "%"+string(n), "invalid directive")
```

Comment:
if you can make an error: 
var errIInvalidDirective = errors.New("invalid directive")
and then in the fmt.Errorf, wrap it with %w, then your tests are easier to write.